### ISSUE TITLE
cuda package blindly installs x86_64 regardless of current arch

### ISSUE THREAD

**Issue by trws** [spack/spack#8027]
See subject.

### Expected Result

Either an error during install, or a working install, not a "successful" install of an incompatible binary.

### Actual Result

An incompatible binary.

### Steps to reproduce the issue

```console
$ # log in to a ppc64le system
$ spack install cuda
$ spack install hwloc+cuda # this will fail to load cuda, because it's the wrong arch
...
```

**Comment by citibeth** [comment]
Can you give more information about what DID happen?  What DID get installed after `spack install cuda`?  What does `spack spec cuda` yield?  What is the contents of your `compilers.yaml` file?

**Comment by trws** [comment]
As the subject says @citibeth, it installs the x86_64 version regardless of the target architecture, into the specific architecture directory (ppc64le in this instance).  The yelling is not helpful by the way.

spack path and file info for installed nvcc:

```
$ /usr/gapps/kras/spack/bin/spack find --path cuda
==> 2 installed packages.
-- linux-rhel7-ppc64le / gcc@4.9.3 ------------------------------
    cuda@9.1.85  /usr/gapps/kras/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.9.3/cuda-9.1.85-klohayjxb4gk2uen44c3ibd7dlegbnt4
$ file /usr/gapps/kras/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.9.3/cuda-9.1.85-klohayjxb4gk2uen44c3ibd7dlegbnt4/bin/nvcc
/usr/gapps/kras/spack/opt/spack/linux-rhel7-ppc64le/gcc-4.9.3/cuda-9.1.85-klohayjxb4gk2uen44c3ibd7dlegbnt4/bin/nvcc: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.4.0, stripped
```

spec cuda:

```
$ /usr/gapps/kras/spack/bin/spack spec cuda
Input spec
--------------------------------
cuda

Concretized
--------------------------------
cuda@9.1.85%gcc@4.9.3 arch=linux-rhel7-ppc64le
```

compilers.yaml:

```yaml
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/bin/clang
      cxx: /usr/tcetmp/bin/clang++
      f77: null
      fc: null
    spec: clang@3.8.0
    target: ppc64le
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/clang/clang-3.9.1/bin/clang
      cxx: /usr/tcetmp/packages/clang/clang-3.9.1/bin/clang++
      f77: null
      fc: null
    spec: clang@3.9.1
    target: ppc64le
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/clang/clang-4.0.0/bin/clang
      cxx: /usr/tcetmp/packages/clang/clang-4.0.0/bin/clang++
      f77: null
      fc: null
    spec: clang@4.0.0
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/clang/clang-coral-2017.03.15/bin/clang
      cxx: /usr/tcetmp/packages/clang/clang-coral-2017.03.15/bin/clang++
      f77: null
      fc: null
    spec: clang@coral-2017.03.15
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/clang/clang-coral-2017.03.29/bin/clang
      cxx: /usr/tcetmp/packages/clang/clang-coral-2017.03.29/bin/clang++
      f77: null
      fc: null
    spec: clang@coral-2017.03.29
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/clang/clang-coral-2017.05.03/bin/clang
      cxx: /usr/tcetmp/packages/clang/clang-coral-2017.05.03/bin/clang++
      f77: null
      fc: null
    spec: clang@coral-2017.05.03
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.8.5
    target: ppc64le

#- compiler:
#    environment: {}
#    extra_rpaths: []
#    flags: {}
#    modules: []
#    operating_system: rhel7
#    paths:
#      cc: /usr/tcetmp/packages/gcc/gcc-4.8-redhat/bin/gcc
#      cxx: /usr/tcetmp/packages/gcc/gcc-4.8-redhat/bin/g++
#      f77: /usr/tcetmp/packages/gcc/gcc-4.8-redhat/bin/gfortran
#      fc: /usr/tcetmp/packages/gcc/gcc-4.8-redhat/bin/gfortran
#    spec: gcc@4.8.5-redhat
#    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/bin/gcc
      cxx: /usr/tcetmp/bin/g++
      f77: /usr/tcetmp/bin/gfortran
      fc: /usr/tcetmp/bin/gfortran
    spec: gcc@4.9.3
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/pgi/pgi-16.10/bin/pgcc
      cxx: /usr/tcetmp/packages/pgi/pgi-16.10/bin/pgc++
      f77: /usr/tcetmp/packages/pgi/pgi-16.10/bin/pgfortran
      fc: /usr/tcetmp/packages/pgi/pgi-16.10/bin/pgfortran
    spec: pgi@16.10
    target: ppc64le
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/pgi/pgi-17.1/bin/pgcc
      cxx: /usr/tcetmp/packages/pgi/pgi-17.1/bin/pgc++
      f77: /usr/tcetmp/packages/pgi/pgi-17.1/bin/pgfortran
      fc: /usr/tcetmp/packages/pgi/pgi-17.1/bin/pgfortran
    spec: pgi@17.1
    target: ppc64le
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/pgi/pgi-17.3/bin/pgcc
      cxx: /usr/tcetmp/packages/pgi/pgi-17.3/bin/pgc++
      f77: /usr/tcetmp/packages/pgi/pgi-17.3/bin/pgfortran
      fc: /usr/tcetmp/packages/pgi/pgi-17.3/bin/pgfortran
    spec: pgi@17.3
    target: ppc64le
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/packages/pgi/pgi-17.4/bin/pgcc
      cxx: /usr/tcetmp/packages/pgi/pgi-17.4/bin/pgc++
      f77: /usr/tcetmp/packages/pgi/pgi-17.4/bin/pgfortran
      fc: /usr/tcetmp/packages/pgi/pgi-17.4/bin/pgfortran
    spec: pgi@17.4
    target: ppc64le

#- compiler:
#    environment: {}
#    extra_rpaths: []
#    flags:
#      fflags: -qzerosize
#    modules: []
#    operating_system: rhel7
#    paths:
#      cc: /usr/tcetmp/bin/xlc-gpu
#      cxx: /usr/tcetmp/bin/xlC-gpu
#      f77: /usr/tcetmp/bin/xlf-gpu
#      fc: /usr/tcetmp/bin/xlf2008-gpu
#    spec: xl@14.1
#    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags:
      fflags: -qzerosize
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/bin/xlc
      cxx: /usr/tcetmp/bin/xlC
      f77: /usr/tcetmp/bin/xlf
      fc: /usr/tcetmp/bin/xlf2008
    spec: xl@14.1
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags:
      fflags: -qzerosize
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/bin/xlc
      cxx: /usr/bin/xlC
      f77: /usr/bin/xlf
      fc: /usr/bin/xlf2008
    spec: xl@13.1
    target: ppc64le

- compiler:
    environment: {}
    extra_rpaths: []
    flags:
      fflags: -qzerosize
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/tcetmp/bin/xlc_r
      cxx: /usr/tcetmp/bin/xlc++_r
      f77: /usr/tcetmp/bin/xlf_r
      fc: /usr/tcetmp/bin/xlf2008_r
    spec: xl_r@14.1
    target: ppc64le
- compiler:
    environment: {}
    extra_rpaths: []
    flags:
      fflags: -qzerosize
    modules: []
    operating_system: rhel7
    paths:
      cc: /usr/bin/xlc_r
      cxx: /usr/bin/xlc++_r
      f77: /usr/bin/xlf_r
      fc: /usr/bin/xlf2008_r
    spec: xl_r@13.1
    target: ppc64le

```

**Comment by LudovicEnault** [comment]
Hi all,

In case it helps, a workaround is to install CUDA manually and add it to packages.yaml:
```
packages:
  cuda:
    buildable: False
    paths:
      cuda@9.2: /products/cuda/9.2
      cuda@system: /usr/local/cuda
    version: [system]
```
Hope it helps

**Comment by alalazo** [comment]
A workaround that solves this issue has been implemented in #13912 (a more stable solution is in the roadmap). Closing.
